### DISCUSSION TITLE
Morph relationship with composite key, is possible?

### DISCUSSION THREAD

**Discussion by underdpt** [staudenmeir/eloquent-has-many-deep#237] [Q&A]
Hello,

So, I have this relationship: an order line **for a product**  belongs to an order, which has many delivery notes, which have many lines **for that product** (and this is a morphed relationship).

The key here is that the order line has two keys, the order id and the product id, and later on the relationship chain, I must filter by that product id, but on that point the relationship is a morphed one.

I've been able to get all the lines for the order from the order line, with this:

```php
public function deliveryLines(): HasManyDeep
{
    return $this->hasManyDeep(
        DeliveryLines::class,
        [Order::class, Deliveries::class],
        [
            'id', 'orders_id', ['documento_type', 'documento_id'],
        ],
        ['orders_id']
    );
}
```

So, the question is... Can this be done?

Thanks!

**Comment by staudenmeir** [comment]
Hi @underdpt,
Please share the schema of all database tables that are involved in this relationship (you can remove the columns that aren't relevant to this question).

**Reply by underdpt** [reply]
Sure, thank you!

Here they are (ordered by the relationship order: order_lines -> orders -> delivery_notes -> delivery_notes_lines):

```sql
CREATE TABLE `order_lines` (
  `id` bigint UNSIGNED NOT NULL,
  `order_id` bigint UNSIGNED NOT NULL,
  `product_id` bigint UNSIGNED NOT NULL,
);

CREATE TABLE `orders` (
  `id` bigint UNSIGNED NOT NULL,
);

CREATE TABLE `delivery_notes` (
  `id` bigint UNSIGNED NOT NULL,
  `order_id` bigint UNSIGNED NOT NULL,
);

CREATE TABLE `delivery_notes_lines` (
  `id` bigint UNSIGNED NOT NULL,
  `product_id` bigint UNSIGNED NOT NULL,
  `documento_type` varchar(125) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `documento_id` bigint UNSIGNED DEFAULT NULL,
);
```

**Reply by staudenmeir** [reply]
Unfortunately, it's not possible to add the `product_id` constraint to the relationship. Composite keys can only be defined between "neighboring" tables.

You can add the constraint to the query when you use the relationships for lazy loading:

```php
$orderLine = OrderLine::find($id);

$orderLine->deliveryLines()->where('product_id')->get();
```

This doesn't work with eager loading and I don't see a feasible way to make it work.

**Reply by underdpt** [reply]
Yes, that's what I though.

Thank you for taking a look!
